apiserver/facades/agent/upgrader/upgrader.go

```diff
@@ -91,33 +95,56 @@ func NewUpgraderAPI(
 }
 
 // WatchAPIVersion starts a watcher to track if there is a new version
-// of the API that we want to upgrade a machine to.
+// of the API that we want to upgrade an application, machine, model or
+// unit to.
 func (u *UpgraderAPI) WatchAPIVersion(ctx context.Context, args params.Entities) (params.NotifyWatchResults, error) {
 	result := params.NotifyWatchResults{
 		Results: make([]params.NotifyWatchResult, len(args.Entities)),
 	}
 	for i, agent := range args.Entities {
-		tag, err := names.ParseMachineTag(agent.Tag)
+		tag, err := names.ParseTag(agent.Tag)
 		if err != nil {
-			return params.NotifyWatchResults{}, errors.Trace(err)
+			result.Results[i].Error = apiservererrors.ParamsErrorf(
+				params.CodeTagInvalid, "%s", err,
+			)
+			continue
 		}
+
 		tagID := tag.Id()
-		upgraderAPIWatcher, err := u.modelAgentService.WatchMachineTargetAgentVersion(ctx, machine.Name(tagID))
+		var errMessage string
+		var upgraderAPIWatcher watcher.NotifyWatcher
+
+		switch tag.Kind() {
+		case names.ControllerTagKind, names.ModelTagKind:
+			errMessage = fmt.Sprintf("model %q", tagID)
+			upgraderAPIWatcher, err = u.modelAgentService.WatchModelTargetAgentVersion(ctx)
+		case names.ApplicationTagKind:
+			errMessage = fmt.Sprintf("application %q", tagID)
+			upgraderAPIWatcher, err = u.modelAgentService.WatchApplicationTargetAgentVersion(ctx, tagID)
+		case names.MachineTagKind:
+			errMessage = fmt.Sprintf("machine %q", tagID)
+			upgraderAPIWatcher, err = u.modelAgentService.WatchMachineTargetAgentVersion(ctx, machine.Name(tagID))
+		case names.UnitTagKind:
+			// Used for non sidecar applications in a kubernetes model.
+			errMessage = fmt.Sprintf("unit %q", tagID)
+			upgraderAPIWatcher, err = u.modelAgentService.WatchUnitTargetAgentVersion(ctx, tagID)
+		default:
+			err = errors.NotValidf("%s", tag.String())
+		}
 
 		switch {
-		case errors.Is(err, errors.NotValid):
+		case errors.Is(err, modelerrors.NotFound), errors.Is(err, applicationerrors.ApplicationNotFound),
+			errors.Is(err, machineerrors.MachineNotFound):
 			result.Results[i].Error = apiservererrors.ParamsErrorf(
-				params.CodeMachineInvalidID,
-				"invalid machine ID %q",
-				tagID,
+				params.CodeNotFound, "%s", errMessage,
 			)
 			continue
 		case err != nil:
 			// We don't understand this error. At this stage we consider it an
 			// internal server error and bail out of the call completely.
 			return params.NotifyWatchResults{}, fmt.Errorf(
-				"cannot watch api version for machine %q: %w",
-				tagID, err,
+				"cannot watch api version for %s: %w",
+				errMessage, err,
```

Comment:
instead of `errMessage` I think all we need to do here is `(cannot watch api version for tag %q: %w", tag, err)`. The stringer interface on the tag will output the kind and id.